streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -220,11 +237,27 @@ public void setStateListener(final KafkaStreams.StateListener listener) {
         stateListener = listener;
     }
 
+    /**
+     * Sets the state
+     * @param newState New state
+     */
     private void setState(final State newState) {
+
         synchronized (stateLock) {
+
+            // there are cases when we shouldn't check if a transition is valid, e.g.,
+            // when, for testing, Kafka Streams is closed multiple times. We could either
+            // check here and immediately return for those cases, or add them to the transition
+            // diagram (but then the diagram would be confusing and have transitions like
+            // NOT_RUNNING->NOT_RUNNING).
+            if (newState != NOT_RUNNING && (state == State.NOT_RUNNING || state == PENDING_SHUTDOWN)) {
```

Comment:
A simple case here is given in the `testCloseIsIdempotent` test where we try to close multiple times. We could put the logic elsewhere, but this attempts to put it all in one place.